backend/dataall/base/api/graphql_api.py

```diff
@@ -0,0 +1,183 @@
+    if issubclass(arg_type, gql.ObjectType) or issubclass(arg_type, gql.InputType):
+        return gql.Ref(_objects[arg_type])
+    return _scalar_serialize_mapping.get(arg_type)
+
+
+def _has_default_value(cls, field_name):
+    for field_info in fields(cls):
+        if field_info.name == field_name:
+            return field_info.default is not field_info.default_factory
+    return False
+
+
+def _create_custom_object(cls, data: dict):
+    # TODO: fix the problem of nested objects in the input
+    return cls(**data)
+
+
+def _process_api_func(name, field):
+    def decorator(f):
+        fn, fn_decorator = process_func(f)
+
+        hints = get_type_hints(fn)
+        signature = inspect.signature(fn)
+        num_args = len(signature.parameters)
+
+        if "return" not in hints:
+            raise ValueError(f"Function {fn.__name__} doesn't specify the return type")
+
+        if len(hints) != num_args + 1:
+            raise ValueError(
+                f"Function {fn.__name__} marked as GraphQL API, "
+                f"but don't specify types for all parameters"
+            )
+        gql_args = []
+        return_type = None
+
+        parameters = signature.parameters
+        for hint_name, hint_type in hints.items():
+            gql_type = _resolve_type(hint_type)
+            if not gql_type:
+                raise ValueError(f"Function {fn.__name__}  takes/returns not a GraphQL type!")
+
+            if hint_name != 'return':
+                param = parameters[hint_name]
+                has_default = param.default is not inspect.Parameter.empty
+                gql_args.append(
+                    gql.Argument(
+                        name=hint_name,
+                        type=gql_type if has_default else gql.NonNullableType(gql_type)
+                    )
+                )
+            else:
+                return_type = gql_type
+
+        def decorated(*args, **kwargs):
+            deserialized: dict = {}
+            for param_name, json_data in kwargs.items():
+                param_type = hints[param_name]
+                if param_type not in _scalars:
+                    deserialized[param_name] = _create_custom_object(param_type, json_data)
+                else:
+                    deserialized[param_name] = json_data
+            return fn(*args, **deserialized)
+
+        field(
+            name=name,
+            args=gql_args if gql_args else None,
+            type=return_type,
+            resolver=fn_decorator(decorated),
+            api_version=2,
+        )
+
+        return fn_decorator(decorated)
+    return decorator
+
+
+def _process_api_class(name, object_factory):
+    if name in _object_names:
+        raise ValueError(f"The object {name} already exists in the schema")
+    _object_names.add(name)
+
+    def class_decorator(cls):
+        _objects[cls] = name
+
+        hints = get_type_hints(cls)
+        gql_args = []
+
```

Comment:
Since input and objects are dataclasses we can assume that their fields have type annotations, but we still need to check the types

Reply:
Yes. The input type and the object fields should also have types. If it doesn't have a type/hint we don't know how to map it to the gql type